File: backend/domain/user/repository.go
package user

import (
	"context"
)

type UserRepository interface {
	// FindByUserId : 指定されたuserIdに基づいてユーザーを取得
	// userId: ユーザーID
	// return: ユーザー情報、またはnil
	FindUserByUserId(ctx context.Context, userId string) (*User, error)

	// FindUserByEmail : 指定されたemailに基づいてユーザーを取得
	// email: メールアドレス
	// return: ユーザー情報、またはnil
	FindUserByEmail(ctx context.Context, email Email) (*User, error)

	// CreateUser : 新しいユーザーを作成
	// userName: ユーザー名
	// email: メールアドレス
	// return: 作成されたユーザー情報
	CreateUser(ctx context.Context, user string, email Email) (*User, error)

	// UpdateUser : ユーザー情報を更新
	// userName: 新しいユーザー名
	// email: 新しいメールアドレス
	// return: 更新されたユーザー情報
	UpdateUser(ctx context.Context, userId string, userName string, email Email) (*User, error)

	// DeleteUser : 指定されたuserIdのユーザーを削除
	// userId: ユーザーID
	// return: 削除の成功状態
	DeleteUser(ctx context.Context, userId string) error
}

File: frontend/src/domain/user/repository.ts
import type { User } from "./entity";
import type { Email } from "./value-object";

/**
 * 指定されたuserIdに基づいてユーザーを取得
 * @param userId - ユーザーID
 * @returns ユーザー情報、またはnull
 */
export type FindUserByUserId = (userId: string) => Promise<User | null>;

/**
 * 指定されたemailに基づいてユーザーを取得
 * @param email - メールアドレス
 * @returns ユーザー情報、またはnull
 */
export type FindUserByEmail = (email: Email) => Promise<User | null>;

/**
 * 新しいユーザーを作成
 * @param userName - ユーザー名
 * @param email - メールアドレス
 * @returns 作成されたユーザー情報
 */
export type CreateUser = (
	userName: string,
	email: Email,
) => Promise<User | null>;

/**
 * ユーザー情報を更新
 * @param userName - 新しいユーザー名
 * @param email - メールアドレス
 * @returns 更新されたユーザー情報
 */
export type UpdateUser = (
	userId: string,
	userName: string,
	email: Email,
) => Promise<User | null>;

/**
 * 指定されたuserIdのユーザーを削除
 * @param userId - ユーザーID
 * @returns 削除の成功状態
 */
export type DeleteUser = (userId: string) => Promise<void>;
